ProcessLib/Parameter/Parameter.cpp

```diff
@@ -0,0 +1,59 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include "Parameter.h"
+#include "BaseLib/ConfigTree.h"
+#include "BaseLib/Error.h"
+
+#include "ConstantParameter.h"
+#include "MeshElementParameter.h"
+#include "MeshNodeParameter.h"
+
+namespace ProcessLib
+{
+
+std::unique_ptr<ParameterBase> createParameter(
+    BaseLib::ConfigTree const& config, std::vector<MeshLib::Mesh*> const& meshes)
+{
+
+    //! \ogs_file_param{parameter__name}
+    auto const name = config.getConfigParameter<std::string>("name");
+    //! \ogs_file_param{parameter__type}
+    auto const type = config.peekConfigParameter<std::string>("type");
+
+    // Create parameter based on the provided type.
+    if (type == "Constant")
+    {
+        INFO("ConstantParameter: %s", name.c_str());
+        auto param = createConstantParameter(config);
+        param->name = name;
+        return param;
+    }
+    else if (type == "MeshElement")
+    {
+        INFO("MeshElementParameter: %s", name.c_str());
+        auto param = createMeshElementParameter(config, *meshes.front());
+        param->name = name;
+        return param;
```

Comment:
A return-by-value is an r-value automatically, so your second suggestion does not make sense.
The first suggestion looks more interesting, but in the past you (or your clang) told me that `return std::move(...)` defeats compiler optimizations.
Since this PR apparently builds (or built for some older state) on all of our platforms, obviously no compiler tries to copy the `unique_ptr`, so I wouldn't investigate it further now.